### PR TITLE
Emulator Runtime Patches

### PR DESCRIPTION
**Pull request by abeisgoat**
Fixes for #1262, #1261, and #1258

### CODE REVIEW

File: src/emulator/functionsEmulatorRuntime.ts [hunk 1]
```diff
@@ -17,11 +17,13 @@ import { spawnSync } from "child_process";
 import * as path from "path";
 import * as admin from "firebase-admin";
 
-function slowRequireResolve(moduleName: string, opts?: { paths: string[] }): string {
-  opts = opts || { paths: [] };
+let app: admin.app.App;
+let adminModuleProxy: typeof admin;
+
+function slowRequireResolve(moduleName: string, cwd?: string): string {
```
Comment (samtstern) [inline src/emulator/functionsEmulatorRuntime.ts]: Can you just add a comment to this method explaining:
  * Why it's "slow"
  * When you should use it

File: src/emulator/functionsEmulatorRuntime.ts [hunk 2]
```diff
@@ -530,40 +535,44 @@ async function ProcessHTTPS(frb: FunctionsRuntimeBundle, trigger: EmulatedTrigge
 
 async function ProcessBackground(
   frb: FunctionsRuntimeBundle,
-  stubbedAdminModule: typeof admin,
   trigger: EmulatedTrigger
 ): Promise<void> {
   const { Change } = require("firebase-functions");
   new EmulatorLog("SYSTEM", "runtime-status", "ready").log();
 
+  if (!app) {
```
Comment (samtstern) [inline src/emulator/functionsEmulatorRuntime.ts]: Should this also be in `ProcessHTTPS`?  Maybe it should be one level higher and just be in `main` before we invoke anything?

File: src/emulator/functionsEmulatorRuntime.ts [hunk 3]
```diff
@@ -530,40 +535,44 @@ async function ProcessHTTPS(frb: FunctionsRuntimeBundle, trigger: EmulatedTrigge
...
+  We use an internal Firestore method makeFirestoreSnapshot to generate Snapshots which we pass to firebase-function's
   Change object. If we have a value for new / old snap, then we create a valid snapshot, otherwise we
-  invoke snapshot_ with a different signature describe here...
+  invoke makeFirestoreSnapshot with a different signature describe here...
   https://github.com/googleapis/nodejs-firestore/blob/114de25d1af3fc7441da3242f6bc8cc8354ffa09/dev/test/index.ts#L574
   To create a snapshot where .exists() fails.
    */
 
   let newSnap;
   if (proto.data.value) {
```
Comment (samtstern) [inline src/emulator/functionsEmulatorRuntime.ts]: Probably out of scope for this PR but I think this block, stubbing the snapshot making, is what was making it hard for me to do callable functions with Firestore.

I found myself having to stuff a `name` into `data.value` and `data.oldValue` and I didn't see any code like that in the old functions emu.

File: src/emulator/types.ts [hunk 4]
```diff
@@ -117,14 +117,18 @@ export class EmulatorLog {
...
 
-  toString(): string {
-    return JSON.stringify({
-      timestamp: this.timestamp,
-      level: this.level,
-      text: this.text,
-      data: this.data,
-      type: this.type,
-    });
+  toString(pretty = false): string {
```
Comment (samtstern) [inline src/emulator/types.ts]: Maybe just two methods: toString() and toPrettyString()

This is a total nit but I find arguments to `toString()` to be pretty strange.

File: src/test/emulators/functionsEmulatorRuntime.spec.ts [hunk 5]
```diff
@@ -197,11 +198,12 @@ describe("FunctionsEmulatorRuntime", () => {
 
       it("should provide non-stubbed non-default app from initializeApp", async () => {
         const serializedTriggers = (() => {
+          require("firebase-admin").initializeApp({}, "non-default");
           return {
             function_id: require("firebase-functions")
               .firestore.document("test/test")
               .onCreate(async () => {
-                require("firebase-admin").initializeApp({}, "non-default");
+                /* */
```
Comment (samtstern) [inline src/test/emulators/functionsEmulatorRuntime.spec.ts]: nit: What's with this empty block comment?

File: src/test/emulators/functionsEmulatorRuntime.spec.ts [hunk 6]
```diff
@@ -213,8 +215,33 @@ describe("FunctionsEmulatorRuntime", () => {
         expect(logs["non-default-admin-app-used"]).to.eq(1);
       }).timeout(TIMEOUT_MED);
 
+      it("should alert when the app is not initialized", async () => {
+        const serializedTriggers = (() => {
+          return {
+            function_id: require("firebase-functions")
+              .firestore.document("test/test")
+              .onCreate(async () => {
+                /* tslint:disable:no-console */
```
Comment (samtstern) [inline src/test/emulators/functionsEmulatorRuntime.spec.ts]: nit: I think this `tslint` comment does nothing.

File: src/test/emulators/functionsEmulatorRuntime.spec.ts [hunk 7]
```diff
@@ -323,11 +351,54 @@ describe("FunctionsEmulatorRuntime", () => {
 
         await runtime.exit;
       }).timeout(TIMEOUT_MED);
+
+      // it("should merge .initializeApp arguments from user", async () => {
```
Comment (samtstern) [inline src/test/emulators/functionsEmulatorRuntime.spec.ts]: Travis guarantees this environment:
```
CI=true
TRAVIS=true
CONTINUOUS_INTEGRATION=true
```

Either deleted the commented out code or prepend a travis check.

**Comment by malekpour** [comment]
@abeisgoat, any idea how soon you are going to merge this and publish the patch to npm?

**Comment by abeisgoat** [comment]
@malekpour Likely today.

**Pull request merged**